Handle unknown dpId in _on_device_report

magior opened the pull request:
Some devices send updates that include `dpId` values not defined in their `local_strategy`.  
Previously this raised a `KeyError`, preventing valid values from being applied.  

This change adds a guard for unmapped `dpId` values:  
- unknown `dpId`s are skipped (logged as a warning),  
- valid datapoints continue to update `device.status`.

Closes #38

magior committed:
Handle unknown dpId in _on_device_report

Skip unmapped dpIds instead of raising KeyError, so valid updates are still applied.

epenet commented on tuya_sharing/manager.py:
```diff
@@ -158,18 +158,21 @@ def _on_device_report(self, device_id: str, status: list):
-                    updated_status_properties.append(code)
-                    if t := item.get("t"):
-                        dp_timestamps[code] = t
+                    if item["dpId"] in device.local_strategy:
```
I think it's easier if you use a `continue` structure with an early exit:
```python
                    if item["dpId"] not in device.local_strategy:
                        logger.debug(f"mq _on_device_report unknown dpId: {item['dpId']}")
                        continue
                    dp_id_item = device.local_strategy[item["dpId"]]
                    strategy_name = dp_id_item["value_convert"]
                    ...
```

Also, I think it's better to use `debug` statement not `warning`, to avoid poluting logs if it's done on purpose and no-one can do anything about it.

magior replied:
Thanks for the feedback. Personally I prefer not to use `continue` as I find it less clear, but it can of course be changed to that style during merge if preferred. 
The same applies to the log level — no problem if it’s switched to `debug`.

FredrikM97 replied:
I agree with @epenet. Using continue here is more appropriate. No point in wrapping everything for 1 line of logging. Also using debug is much better.

epenet commented:
@shihuimiao  / @zzysszzy can you please take a look at this?
I think it will solve a lot of issues for Home Assistant users missing data

epenet commented:
@shihuimiao / @zzysszzy is there any feedback on this?

epenet commented on tuya_sharing/manager.py:
```diff
@@ -158,18 +158,21 @@ def _on_device_report(self, device_id: str, status: list):
+                        if t := item.get("t"):
+                            dp_timestamps[code] = t
+                    else:
+                        logger.warning(f"mq _on_device_report unknown dpId: {item['dpId']}")
```
Note: I really think a warning is too much.

```suggestion
                        logger.debug(f"mq _on_device_report unknown dpId: {item['dpId']}")
```

epenet commented:
@shihuimiao / @zzysszzy what can we do to get a review of this?

epenet commented:
Log from @PedroDelCargo
```
2025-09-30 11:56:24.500 DEBUG (paho-mqtt-client-cloud_db23a80dd84937362c33d8a3c0bcbf7b) [tuya_sharing] _on_log: Received PUBLISH (d0, q0, r0, m0), 'cloud/device/bfe251cfa8882fae3cpvup/in/24d2afb98121974b6505c40a3a980362/pen', ...  (370 bytes)
2025-09-30 11:56:24.501 DEBUG (paho-mqtt-client-cloud_db23a80dd84937362c33d8a3c0bcbf7b) [tuya_sharing] payload-> b'{"protocol":4,"data":{"devId":"bfe251cfa8882fae3cpvup","dataId":"30b03adb-93b5-4b24-8f56-19e6d1dc7eb5","productKey":"rknwi0ctbbghzgla","status":[{"dpId":101,"t":1759226184088,"value":456},{"dpId":3,"t":1759226184088,"value":87},{"dpId":15,"t":1759226184088,"value":40},{"dpId":5,"t":1759226184088,"value":76},{"dpId":9,"t":1759226184088,"value":"c"}]},"t":1759226184484}'
2025-09-30 11:56:24.501 DEBUG (paho-mqtt-client-cloud_db23a80dd84937362c33d8a3c0bcbf7b) [tuya_sharing] on_message: {'protocol': 4, 'data': {'devId': 'bfe251cfa8882fae3cpvup', 'dataId': '30b03adb-93b5-4b24-8f56-19e6d1dc7eb5', 'productKey': 'rknwi0ctbbghzgla', 'status': [{'dpId': 101, 't': 1759226184088, 'value': 456}, {'dpId': 3, 't': 1759226184088, 'value': 87}, {'dpId': 15, 't': 1759226184088, 'value': 40}, {'dpId': 5, 't': 1759226184088, 'value': 76}, {'dpId': 9, 't': 1759226184088, 'value': 'c'}]}, 't': 1759226184484}
2025-09-30 11:56:24.501 DEBUG (paho-mqtt-client-cloud_db23a80dd84937362c33d8a3c0bcbf7b) [tuya_sharing] mq receive-> {'protocol': 4, 'data': {'devId': 'bfe251cfa8882fae3cpvup', 'dataId': '30b03adb-93b5-4b24-8f56-19e6d1dc7eb5', 'productKey': 'rknwi0ctbbghzgla', 'status': [{'dpId': 101, 't': 1759226184088, 'value': 456}, {'dpId': 3, 't': 1759226184088, 'value': 87}, {'dpId': 15, 't': 1759226184088, 'value': 40}, {'dpId': 5, 't': 1759226184088, 'value': 76}, {'dpId': 9, 't': 1759226184088, 'value': 'c'}]}, 't': 1759226184484}
2025-09-30 11:56:24.501 DEBUG (paho-mqtt-client-cloud_db23a80dd84937362c33d8a3c0bcbf7b) [tuya_sharing] mq _on_device_report-> [{'dpId': 101, 't': 1759226184088, 'value': 456}, {'dpId': 3, 't': 1759226184088, 'value': 87}, {'dpId': 15, 't': 1759226184088, 'value': 40}, {'dpId': 5, 't': 1759226184088, 'value': 76}, {'dpId': 9, 't': 1759226184088, 'value': 'c'}]
2025-09-30 11:56:24.501 ERROR (paho-mqtt-client-cloud_db23a80dd84937362c33d8a3c0bcbf7b) [tuya_sharing] on message error = 101
```

As you can see, we get `mq _on_device_report->` but never `mq _on_device_report before strategy convert strategy_name`
Instead, we get `on message error`

epenet commented:
@shihuimiao / @zzysszzy what can we do to get a review of this?

magior committed:
Refactor _on_device_report:

- replace nested if/else with continue for unknown dpId
- change log level from warning to debug

magior commented:
I've updated the PR with the suggested changes.
Hope this can get a review soon — it should fix issues affecting many devices on Home Assistant.

Mailootje commented:
Hi, as I'm the OP of this issue:

https://github.com/home-assistant/core/issues/155064

Many devices won't update their status in HomeAssistant. Can you please fix this issue/merge this PR?

shihuimiao commented:
> @shihuimiao / @zzysszzy is there any feedback on this?

ok I will handle this issue as soon as possible.

makol1995 commented:
> Hi, as I'm the OP of this issue:
> 
> [home-assistant/core#155064](https://github.com/home-assistant/core/issues/155064)
> 
> Many devices won't update their status in HomeAssistant. Can you please fix this issue/merge this PR?

I have the same issue.

shihuimiao commented:
> @shihuimiao / @zzysszzy is there any feedback on this?

https://pypi.org/project/tuya-device-sharing-sdk/0.2.5/
Already uploaded

epenet commented:
Thanks @shihuimiao !
Can you please also take a look at #37 ?

shihuimiao commented:
> Thanks @shihuimiao ! Can you please also take a look at #37 ?

ok

JimboFarns commented:
Sorry - Maybe I am missing something - but can someone help or confirm how I can get my Tuya integration from 0.2.4 to 0.2.5?

PedroDelCargo commented:
> Sorry - Maybe I am missing something - but can someone help or confirm how I can get my Tuya integration from 0.2.4 to 0.2.5?

Usually, Tuya integration updates come with HA updates.
HA 2025.11.1 solved issues I had with Tuya devices.

epenet commented:
> Sorry - Maybe I am missing something - but can someone help or confirm how I can get my Tuya integration from 0.2.4 to 0.2.5?

It was automatically included in HA `2025.11.1`
See https://github.com/home-assistant/core/pull/156014

If you are not using Home Assistant, or you are using a custom component, then I am not sure.
